python/ray/serve/controller.py

```diff
@@ -308,12 +309,21 @@ async def kill_actor(replica_name_to_use):
                     except ValueError:
                         return
 
-                    # TODO(edoakes): this logic isn't ideal because there may
-                    # be pending tasks still executing on the replica. However,
-                    # if we use replica.__ray_terminate__, we may send it while
-                    # the replica is being restarted and there's no way to tell
-                    # if it successfully killed the worker or not.
-                    ray.kill(replica, no_restart=True)
+                    try:
+                        await asyncio.wait_for(
+                            replica.shutdown.remote(),
+                            timeout=DEFAULT_GRACEFUL_SHUTDOWN_CONTROLLER_WAIT_S
+                        )
+                    except ray.exceptions.RayActorError:
+                        # Actor should have exited.
+                        return
+                    except asyncio.TimeoutError:
+                        # Graceful period passed, kill it forcefully.
+                        logger.debug(
+                            f"{replica_name_to_use} did not shutdown after "
+                            "{DEFAULT_GRACEFUL_SHUTDOWN_CONTROLLER_WAIT_S}, "
+                            "killing.")
+                        ray.kill(replica, no_restart=True)
```

Comment:
Once we have the asynchronous control loop we can handle this in a better way. Is there any reason not to make this timeout much longer though? In most cases this codepath should never be reached.